### ISSUE TITLE
Disable container discovery for kubernetes_sd_config endpoints

### ISSUE THREAD

**Issue by m-messiah** [prometheus/prometheus#9357]
## Proposal

> The endpoints role discovers targets from listed endpoints of a service. For each endpoint address one target is discovered per port. If the endpoint is backed by a pod, all additional container ports of the pod, not bound to an endpoint port, are discovered as targets as well.

Sometimes you don't want to have such discovery, in addition, to already existent endpoints and it would be nice to have the ability to disable such unused but **heavy** behaviour.
Such behaviour may break Kubernetes API by tons of requests - when Prometheus gets endpoint discovery and understand it is the pod behind - it sends a LIST pods request to the API. And this happens for each `endpoints` object, which leads to apiserver heavy load in big clusters.

**Comment by brancz** [comment]
Do you have a reproducible setup for this (maybe with audit logs)? The way I read the code is that we list/watch endpoints and pods only once and do all the following operations on the informer cache, so this should not be causing any more load on the Kubernetes API than the initial list/watch (once for endpoints once for pods). Prometheus also re-uses list/watches as best as possible. For example two kubernetes service discoveries that are identical except in their relabel configs also share a list-watch. The only pathological case that I am aware of is when there is a kubernetes service discovery per namespace, as then a list-watch per namespace is needed, if this is the case you are running into, then I would suggest instead of using the `namespaces` field in the kubernetes service discovery config, use relabeling to perform namespace selection instead (as that way the list-watch can be shared between configs).

**Comment by m-messiah** [comment]
Yes, we have per-namepace service monitors  and I am really worried about using a shared list-watch and if this request will be too heavy for apiserver

**Comment by m-messiah** [comment]
Actually, I played with it and got almost empty targets lists, because the requests were too heavy, so timed out or killed by apiserver or Prometheus from time to time. And Prometheus Operator brought its own case too.

**Comment by bboreham** [comment]
Reviewed at the bug-scrub meeting:
We have added [EndpointSlice discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#endpointslice) since then, and no further comments in 3 years; closing.
If there are further problems please open a new issue with full details.
